src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -1648,7 +1678,29 @@ optimize traceSteps ctx env (Qualified mn (Ident id)) initN originalExpr =
         Tuple rewrite expr2
 
 freeze :: BackendExpr -> Tuple BackendAnalysis NeutralExpr
-freeze init = Tuple (analysisOf init) $ foldBackendExpr NeutralExpr (\_ neutExpr -> neutExpr) init
+freeze = freezeWithLetRewrites 0 []
+
+freezeWithLetRewrites :: Int -> Array LetRewrite -> BackendExpr -> Tuple BackendAnalysis NeutralExpr
+freezeWithLetRewrites levelShrinkage rewrites init =
+  Tuple (analysisOf init)
+    $ foldBackendExpr
+        ( case _ of
+            Local maybeIdent level
+              | Just qualifiedIdent <- findMap (\(LetRewrite { oldLevel, oldIdent, newQualifiedIdent }) -> if oldLevel == level && oldIdent == maybeIdent then Just newQualifiedIdent else Nothing) rewrites -> NeutralExpr $ Var qualifiedIdent
+            Local maybeIdent (Level level) -> NeutralExpr $ Local maybeIdent (Level (level - levelShrinkage))
+            Let ident (Level level) binding body -> NeutralExpr $ Let ident (Level (level - levelShrinkage)) binding body
+            LetRec (Level level) bindings body -> NeutralExpr $ LetRec (Level (level - levelShrinkage)) bindings body
+            Abs idents body -> NeutralExpr $ Abs (shrink idents) body
+            EffectBind ident (Level level) binding body -> NeutralExpr $ EffectBind ident (Level (level - levelShrinkage)) binding body
+            UncurriedAbs idents body -> NeutralExpr $ UncurriedAbs (shrink idents) body
+            UncurriedEffectAbs idents body -> NeutralExpr $ UncurriedEffectAbs (shrink idents) body
+            x -> NeutralExpr x
+        )
+        (\_ neutExpr -> neutExpr)
+        init
+  where
+  shrink :: forall f b. Functor f => f (Tuple b Level) -> f (Tuple b Level)
+  shrink = map (\(Tuple i (Level l)) -> Tuple i (Level (l - levelShrinkage)))
```

Comment:
Originall I had suggested that rewriting could be done as part of freeze, but in a recent discord comment, I had suggested potentially:
> In looking at this closer, I still don't think this needs to be part of optimize. I think optimize can stay as-is. It won't be necessary to optimize the bindings again, all you need is correct analysis. Otherwise your suggestion makes sense. I'm understanding this to be something like:
> * Optimize a top-level expression.
> * Check the resulting BackendExpr to see if it's a Let/LetRec.
> * If so, eval (not optimize) the Expr one more time. This will give you a SemLet/SemLetRec which has a continuation for the substitution.
> * Recursively, quote the bound expr, synthesize a new top-level identifier, and toExternImpl it. Call the continuation with the fresh identifier.
>
> This should yield multiple bindings, each with an ExternImpl with correct analysis, and also substitute the names correctly.

It looks like there is quite a lot of work involved in manually doing this rewrite in freeze, much more so than I originally had in mind. There are two key points:

* It's not technically necessary for correctness (as far as I can tell), but adjusting all the levels is something we'd want to do for sanity and consistency.
* We also want to make sure we rebuild correct analysis for everything that's going to go in "implementations".

Both of these things are already taken care of automatically by just going through another eval/quote cycle like in my followup suggestion. It feels like it may be worth it to me to just eat the cost of another pass rather than implement custom level adjustments in freeze. It also feels like it would greatly simplify `toExternImpl`, as that could just stay as-is with an outer loop to collect multiple implementations from one optimized definition. What do you think?